sql/optbuilder/builder.go

```diff
@@ -0,0 +1,1309 @@
+		innerExpr := b.buildScalar(inScope, v.Expr)
+		//TODO: rename this from Charset to Collation
+		collation, err := sql.ParseCollation(nil, &v.Charset, false)
+		if err != nil {
+			b.handleErr(err)
+		}
+		// If we're collating a string literal, we check that the charset and collation match now. Other string sources
+		// (such as from tables) will have their own charset, which we won't know until after the parsing stage.
+		charSet := b.ctx.GetCharacterSet()
+		if _, isLiteral := innerExpr.(*expression.Literal); isLiteral && collation.CharacterSet() != charSet {
+			b.handleErr(sql.ErrCollationInvalidForCharSet.New(collation.Name(), charSet.Name()))
+		}
+		expression.NewCollatedExpression(innerExpr, collation)
+	case *ast.ValuesFuncExpr:
+		col := b.buildScalar(inScope, v.Name)
+		fn, err := b.cat.Function(b.ctx, "values")
+		if err != nil {
+			b.handleErr(err)
+		}
+		values, err := fn.NewInstance([]sql.Expression{col})
+		if err != nil {
+			b.handleErr(err)
+		}
+		return values
+	case *ast.ExistsExpr:
+		selScope := b.buildSelectStmt(inScope, v.Subquery.Select)
+		// rebuild subquery
+		// renameColumns
+		selectString := ast.String(v.Subquery.Select)
+		sq := plan.NewSubquery(selScope.node, selectString)
+		return plan.NewExistsSubquery(sq)
+	case *ast.TimestampFuncExpr:
+		var (
+			unit  sql.Expression
+			expr1 sql.Expression
+			expr2 sql.Expression
+		)
+
+		unit = expression.NewLiteral(v.Unit, types.LongText)
+		expr1 = b.buildScalar(inScope, v.Expr1)
+		expr2 = b.buildScalar(inScope, v.Expr2)
+
+		if v.Name == "timestampdiff" {
+			return function.NewTimestampDiff(unit, expr1, expr2)
+		} else if v.Name == "timestampadd" {
+			return nil
+		}
+		return nil
+	case *ast.ExtractFuncExpr:
+		var unit sql.Expression = expression.NewLiteral(strings.ToUpper(v.Unit), types.LongText)
+		expr := b.buildScalar(inScope, v.Expr)
+		return function.NewExtract(unit, expr)
+	default:
+		b.handleErr(sql.ErrUnsupportedSyntax.New(ast.String(e)))
+	}
+	return nil
+}
+
+func (b *PlanBuilder) buildGetField(inScope *scope, v *ast.ColName) *expression.GetField {
+	table := strings.ToLower(v.Qualifier.String())
+	col := strings.ToLower(v.Name.String())
+	checkScope := inScope
+	outerLen := inScope.outerScopeLen()
+	for checkScope != nil {
+		for i, c := range checkScope.cols {
+			if c.col == col && (c.table == table || table == "") {
+				return expression.NewGetFieldWithTable(outerLen+i, c.typ, c.table, c.col, c.nullable)
+			}
+		}
+		checkScope = checkScope.parent
+	}
+	b.handleErr(sql.ErrColumnNotFound.New(v))
+	return nil
+}
+
+func (b *PlanBuilder) resolveColumn(inScope *scope, v *ast.ColName) (scopeColumn, int) {
+	table := strings.ToLower(v.Qualifier.String())
+	col := strings.ToLower(v.Name.String())
+	for i, c := range inScope.cols {
+		if c.col == col && (c.table == table || table == "") {
+			return c, i
+		}
+	}
+	return scopeColumn{}, -1
+}
+
+func (b *PlanBuilder) resolveOuterColumn(inScope *scope, v *ast.ColName) (scopeColumn, int) {
```

Comment:
Why isn't this recursive? Why are we checking exactly one parent scope instead of following to the root?

For that matter why isn't this logic baked into the regular resolveColumn method